io-engine/src/rebuild/rebuild_log.rs

```diff
@@ -0,0 +1,135 @@
+use bit_vec::BitVec;
+use parking_lot::Mutex;
+use spdk_rs::IoType;
+use std::{
+    fmt::{Debug, Formatter},
+    sync::Arc,
+};
+
+use super::SEGMENT_SIZE;
+
+// Returns ceil of an integer division.
+fn div_ceil(a: u64, b: u64) -> u64 {
+    (a + b - 1) / b
+}
+
+/// Rebuild log data structure.
+pub struct RebuildLog {
+    /// Name of the underlying device, stored for diagnostic purposes.
+    device_name: String,
+    /// Segment bitmap. Ones indicate modified segments.
+    segments: BitVec,
+    /// Device size in segments.
+    num_segments: u64,
+    /// Device size in block.
+    num_blocks: u64,
+    /// Size of block in bytes.
+    block_len: u64,
+}
+
+impl Debug for RebuildLog {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "Rebuild log: '{}' ({} segments: {} blocks x {})",
+            self.device_name,
+            self.num_segments,
+            self.num_blocks,
+            self.block_len
+        )
+    }
+}
+
+impl RebuildLog {
+    /// Creates a new rebuild log instance for the given device.
+    pub fn new(device_name: &str, num_blocks: u64, block_len: u64) -> Self {
+        let num_segments = div_ceil(num_blocks * block_len, SEGMENT_SIZE);
+        let mut segments = BitVec::new();
+        segments.grow(num_segments as usize, false);
+        Self {
+            device_name: device_name.to_owned(),
+            segments,
+            num_segments,
+            num_blocks,
+            block_len,
+        }
+    }
+
+    /// Logs the given operation, marking the corresponding segment as modified
+    /// in the case of a write operation.
+    fn log_op(&mut self, _io_type: IoType, block_offset: u64, num_blocks: u64) {
+        let start_seg = self.segment_index(block_offset);
+        let end_seg = self.segment_index(block_offset + num_blocks);
```

Comment:
should this be like this? Am I missing something?
        let end_seg = self.segment_index(block_offset + num_blocks*block_len);

Reply:
nit: I'd also suggest renaming num_blocks parameter to something like op_blocks_cnt that would indicate number of blocks affected by this op, since there is a field num_blocks already in the RebuildLog that indicates device size as number of  blocks.

Reply:
Ok, I renamed arg names for clarity.